### PR TITLE
[CELEBORN-1070] Add error-prone to pom.xml.

### PR DESCRIPTION
**Pull request by mridulm**
### What changes were proposed in this pull request?

Add [error prone](https://errorprone.info/) to the build.
Error Prone is a static analysis tool that cam catch common bugs and mistakes during compilation.

### Why are the changes needed?
Catch potential issues during build

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Local build

### CODE REVIEW

**Comment by mridulm** [comment]
I was testing it on JDK 11 and later locally - and so did not observe the failures that build system surfaced.
Have added workarounds to fix this, as per error prone documentation.

+CC @waitinfuture

**Comment by SteNicholas** [comment]
@mridulm, could you also test it on JDK17? At present Celeborn supports JDK17.

**Comment by mridulm** [comment]
So looks like there are more failures to be fixed in flink related profiles.
Will convert this back to draft and address those first.

**Comment by waitinfuture** [comment]
> @mridulm, could you also test it on JDK17? At present Celeborn supports JDK17.

Seems GA with JDK 17 fails.

**Comment by mridulm** [comment]
Yes, looks like JDK 17 has some specific instructions as well - did not realize Celeborn supports jdk 17 :) will add those as well, after I fix the flink related failures.

**Comment by pan3793** [comment]
if applicable, would you mind adding such compiling improvement to sbt building system as well?

background: in the previous discussion, we'd like to migrate the building system from maven to sbt because:

- sbt is much faster than maven
- Spark is going to use sbt only in 4.0.0
- sbt shell is convenient for testing during development

the main process is driven by @cfmcgrady, and the first step is to make sbt align with maven(almost reached), and then drop maven ...

**Comment by SteNicholas** [comment]
> Yes, looks like JDK 17 has some specific instructions as well - did not realize Celeborn supports jdk 17 :) will add those as well, after I fix the flink related failures.

@mridulm, the [build](https://github.com/apache/incubator-celeborn#build) section has documented the situation of JDK17 support. You could refer this section.

**Comment by mridulm** [comment]
@pan3793
You mentioned:
> maven scalatest plugin does not support always fork mode

I am currently running with fork mode = true for all the tests and it is running successfully - can you give more details on this please ?

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Comment by waitinfuture** [comment]
> @pan3793 You mentioned:
> 
> > maven scalatest plugin does not support always fork mode
> 
> I am currently running with fork mode = true for all the tests with maven and it is running successfully - can you give more details on this please ? There are so many combinations, I might have missed some !
> 
> > the main process is driven by @cfmcgrady, and the first step is to make sbt align with maven(almost reached), and then drop maven ...
> 
> I did not find sbt support details, can you point me to it ? I would prefer to add the support as a follow up if it is not yet merged.

Here is sbt doc https://celeborn.apache.org/docs/latest/developers/sbt/

**Comment by pan3793** [comment]
> I am currently running with fork mode = true for all the tests with maven and it is running successfully

Yes, we fixed several cases to make it pass now.

Basically, when setting `forkMode=once`, all test cases run on the same forked JVM process, which does not much help for projects that compose a lot of singletons(if we initialize one singleton in the previous test case, the following test cases may be affected, so we must add some reset logic for testing purposes)

The ideal solution is `forkMode=always`, unfortunately, Maven JUnit plugin supports it but scalatest does not.

https://github.com/scalatest/scalatest-maven-plugin/blob/383f396162b7654930758b76a0696d3aa2ce5686/src/main/java/org/scalatest/tools/maven/AbstractScalaTestMojo.java#L162-L168

> I did not find sbt support details ...

Currently, we only provide developer docs with sbt

https://celeborn.apache.org/docs/latest/developers/sbt/

File: pom.xml [hunk 1]
```diff
@@ -864,6 +867,33 @@
           </executions>
         </plugin>
 
+        <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
+          <artifactId>maven-compiler-plugin</artifactId>
+          <version>3.11.0</version>
```
Comment (pan3793) [inline pom.xml:873]: nit: extract as a maven property

File: pom.xml [hunk 2]
```diff
@@ -1128,6 +1171,46 @@
...
+                <path>
+                  <groupId>com.google.errorprone</groupId>
+                  <artifactId>error_prone_core</artifactId>
+                  <version>${error-prone.version}</version>
+                </path>
+                <!-- Other annotation processors go here.
+
+                If 'annotationProcessorPaths' is set, processors will no longer be
+                discovered on the regular -classpath; see also 'Using Error Prone
+                together with other annotation processors' below. -->
```
Comment (pan3793) [inline pom.xml:1208]: > see also 'Using Error Prone together with other annotation processors' below

Emm... what's the below?

Comment (mridulm) [reply pom.xml:1208]: These are coming from the error prone's pom.xml snippets ([1](https://errorprone.info/docs/installation), [2](https://github.com/google/error-prone/blob/f8e33bc460be82ab22256a7ef8b979d7a2cacaba/docs/installation.md#maven)) - looks like they are referencing rest of the doc.
Let me drop the comment, it is not very useful in Celeborn's context

File: pom.xml [hunk 3]
```diff
@@ -1108,6 +1138,19 @@
...
         <java.version>8</java.version>
       </properties>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-compiler-plugin</artifactId>
+            <configuration>
+              <compilerArgs combine.children="append">
+                <arg>-J-Xbootclasspath/p:${settings.localRepository}/com/google/errorprone/javac/${error-prone.jdk8.workaround.javac.version}/javac-${error-prone.jdk8.workaround.javac.version}.jar</arg>
```
Comment (pan3793) [inline pom.xml:1148]: would you mind adding the following link as a comment here?

https://github.com/google/error-prone/blob/f8e33bc460be82ab22256a7ef8b979d7a2cacaba/docs/installation.md#jdk-8


BTW, I think `error-prone.jdk8.javac.version` is clear enough

File: pom.xml [hunk 4]
```diff
@@ -73,6 +73,9 @@
     <commons-lang3.version>3.12.0</commons-lang3.version>
     <commons-io.version>2.13.0</commons-io.version>
     <commons-crypto.version>1.0.0</commons-crypto.version>
+    <!-- last version to support compilation in java 8 -->
```
Comment (SteNicholas) [inline pom.xml:76]: Could this add the link that annotation with last version?

File: pom.xml [hunk 5]
```diff
@@ -1128,6 +1171,46 @@
       <properties>
         <java.version>17</java.version>
       </properties>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-compiler-plugin</artifactId>
+            <version>3.11.0</version>
```
Comment (SteNicholas) [inline pom.xml:1179]: Extract as a maven property too.

Comment (mridulm) [reply pom.xml:1176]: Addressed as part of https://github.com/apache/incubator-celeborn/pull/2025#discussion_r1369491406

File: pom.xml [hunk 6]
```diff
@@ -1128,6 +1168,41 @@
...
```
Comment (SteNicholas) [inline pom.xml:1176]:
```suggestion
            <version>${maven.plugin.compiler.version}</version>
```

**Review by pan3793 — APPROVED** [review]

**Review by SteNicholas — APPROVED** [review]
LGTM. Thanks for @mridulm updates.

**Comment by pan3793** [comment]
Thanks, merged to main for 0.4

**Comment by pan3793** [comment]
@mridulm seems you didn't set up the git email address properly, `mridulatgmail.com`, would you mind replacing the `at` with `@` in your local git email setup? So that the GitHub could correctly link the commit to your GitHub account.

PS: I fixed the email when merging so that the GitHub link works, I saw the unrecognized commits that you merged in the Spark repo.

image

image

**Comment by pan3793** [comment]
alright...

**Pull request closed without merging**